File: June-LeetCoding-Challenge/09-Is-Subsequence/Is-Subsequence.cpp
class Solution1 {
public:
    bool isSubsequence(string s, string t) {
        int i=0, j=0;
        while(i<s.size() && j<t.size()){
            if(t[j] == s[i]) ++j, ++i;
            else ++j;
        }
        if(i == s.size()) return true;
        else return false;
    }
};

class Solution2 {
public:
    bool isSubsequence(string s, string t) {
        int n = s.length();
        int m = t.length();
        vector<vector<bool>> dp(n+1, vector<bool>(m+1, false));
        for (int i = 0; i <= m; i++) dp[0][i] = true;
        for (int i = 1; i <= n; i++)
            for (int j = 1; j <= m; j++){
                if (s[i-1] == t[j-1]) dp[i][j] = dp[i][j] or dp[i-1][j-1];
                dp[i][j] = dp[i][j] or dp[i][j-1];
            }
        return dp[n][m];
    }
};

File: June-LeetCoding-Challenge/09-Is-Subsequence/Is-Subsequence.kt
class IsSubsequenceKotlin392 {

    fun isSubsequence(s: String, t: String): Boolean {
        if (s.isEmpty()) {
            return true
        }
        var indexS = 0
        t.forEach {
            if (it == s[indexS]) {
                if (++indexS == s.length) {
                    return true
                }
            }
        }
        return false
    }

    /*
    fun isSubsequence(s: String, t: String): Boolean {
        val tMap = mutableMapOf<Char, MutableList<Int>>()
        t.forEachIndexed { index, c ->
            tMap.computeIfAbsent(c) { mutableListOf() }.add(index)
        }
        var indexOft = 0
        s.forEach {
            indexOft = binarySearch(tMap[it], indexOft)
            if (indexOft++ == -1) {
                return false
            }
        }
        return true
    }

    private fun binarySearch(nums: List<Int>?, target: Int): Int {
        if (nums.isNullOrEmpty()) {
            return -1
        }
        var left = 0
        var right = nums.size - 1
        while (left + 1 < right) {
            val mid = left + (right - left) / 2
            when {
                nums[mid] <= target -> left = mid
                nums[mid] > target -> right = mid
            }
        }
        return when {
            nums[left] >= target -> nums[left]
            nums[right] >= target -> nums[right]
            else -> -1
        }
    }
     */
}